File: deploy/offline/offline-server.py
#!/usr/bin/env python3
"""Dependency-free local static server for the RaptorQR offline bundle."""

import argparse
import http.server
import mimetypes
import os
import threading
import webbrowser
from pathlib import Path


WEB_ROOT = Path(__file__).resolve().parent / "web"
DEFAULT_PORT = 4173


class OfflineHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=str(WEB_ROOT), **kwargs)

    def end_headers(self):
        self.send_header("Cache-Control", "no-cache")
        self.send_header("X-Content-Type-Options", "nosniff")
        super().end_headers()

    def send_head(self):
        requested = self.translate_path(self.path)
        if not os.path.exists(requested):
            path_without_query = self.path.split("?", 1)[0]
            if not Path(path_without_query).suffix:
                original_path = self.path
                self.path = "/index.html"
                try:
                    return super().send_head()
                finally:
                    self.path = original_path
        return super().send_head()

    def log_message(self, format, *args):
        print(f"[{self.log_date_time_string()}] {format % args}")


def parse_args():
    parser = argparse.ArgumentParser(description="Serve RaptorQR without Internet access")
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=DEFAULT_PORT)
    parser.add_argument("--no-browser", action="store_true")
    return parser.parse_args()


def create_server(host, initial_port):
    for port in range(initial_port, initial_port + 20):
        try:
            return http.server.ThreadingHTTPServer((host, port), OfflineHandler), port
        except OSError:
            continue
    raise RuntimeError(f"No free port found in range {initial_port}-{initial_port + 19}")


def main():
    args = parse_args()
    if not (WEB_ROOT / "index.html").is_file():
        raise SystemExit(f"Offline web files are missing: {WEB_ROOT}")

    mimetypes.add_type("application/wasm", ".wasm")
    server, port = create_server(args.host, args.port)
    local_url = f"http://localhost:{port}/"
    print(f"RaptorQR is available on this computer at {local_url}")
    print("Press Ctrl-C to stop.")

    if not args.no_browser:
        threading.Timer(0.2, lambda: webbrowser.open(local_url)).start()

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nStopping RaptorQR...")
    finally:
        server.server_close()
if __name__ == "__main__":
    main()

File: release/raptorqr-offline/offline-server.mjs
#!/usr/bin/env node
import { createServer } from 'node:http';
import { existsSync, readFileSync, statSync } from 'node:fs';
import { extname, join, normalize, resolve, sep } from 'node:path';
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const root = resolve(fileURLToPath(new URL('./web/', import.meta.url)));
const args = process.argv.slice(2);
const requestedPort = numberArg('--port', 4173);
const host = stringArg('--host', '127.0.0.1');
const noBrowser = args.includes('--no-browser');

if (!existsSync(join(root, 'index.html'))) {
  throw new Error(`Offline web files are missing: ${root}`);
}

const MIME_TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.webmanifest': 'application/manifest+json; charset=utf-8',
  '.wasm': 'application/wasm',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.gif': 'image/gif',
  '.ico': 'image/x-icon',
};

const server = createServer((request, response) => {
  try {
    if (request.method !== 'GET' && request.method !== 'HEAD') {
      sendText(response, 405, 'Method not allowed');
      return;
    }

    const requestUrl = new URL(request.url ?? '/', 'http://localhost');
    const relativePath = decodeURIComponent(requestUrl.pathname).replace(/^\/+/, '');
    let filePath = resolve(root, normalize(relativePath));
    if (filePath !== root && !filePath.startsWith(`${root}${sep}`)) {
      sendText(response, 403, 'Forbidden');
      return;
    }

    if (existsSync(filePath) && statSync(filePath).isDirectory()) {
      filePath = join(filePath, 'index.html');
    }
    if (!existsSync(filePath) && !extname(relativePath)) {
      filePath = join(root, 'index.html');
    }
    if (!existsSync(filePath) || !statSync(filePath).isFile()) {
      sendText(response, 404, 'Not found');
      return;
    }

    const content = readFileSync(filePath);
    response.writeHead(200, {
      'Content-Type': MIME_TYPES[extname(filePath).toLowerCase()] ?? 'application/octet-stream',
      'Content-Length': content.byteLength,
      'Cache-Control': 'no-cache',
      'X-Content-Type-Options': 'nosniff',
    });
    response.end(request.method === 'HEAD' ? undefined : content);
  } catch (error) {
    sendText(response, 500, error instanceof Error ? error.message : String(error));
  }
});

listen(requestedPort);

function listen(port) {
  server.once('error', (error) => {
    if (error.code === 'EADDRINUSE' && port < requestedPort + 19) {
      listen(port + 1);
      return;
    }
    throw error;
  });
  server.listen(port, host, () => {
    const address = server.address();
    const activePort = typeof address === 'object' && address ? address.port : port;
    const url = `http://localhost:${activePort}/`;
    console.log(`RaptorQR is available on this computer at ${url}`);
    console.log('Press Ctrl-C to stop.');
    if (!noBrowser) openBrowser(url);
  });
}

function sendText(response, status, message) {
  const content = Buffer.from(message);
  response.writeHead(status, {
    'Content-Type': 'text/plain; charset=utf-8',
    'Content-Length': content.byteLength,
  });
  response.end(content);
}

function stringArg(name, fallback) {
  const index = args.indexOf(name);
  return index >= 0 && args[index + 1] ? args[index + 1] : fallback;
}

function numberArg(name, fallback) {
  const parsed = Number(stringArg(name, String(fallback)));
  return Number.isInteger(parsed) && parsed > 0 && parsed <= 65535 ? parsed : fallback;
}

function openBrowser(url) {
  const command = process.platform === 'darwin'
    ? ['open', [url]]
    : process.platform === 'win32'
      ? ['cmd', ['/c', 'start', '', url]]
      : ['xdg-open', [url]];
  const child = spawn(command[0], command[1], { detached: true, stdio: 'ignore' });
  child.on('error', () => {});
  child.unref();
}
